### PR TITLE
Use git describe if possible for tagging

### PR DESCRIPTION
**Pull request by tamasgal**
I use annotated Git tags (`git tag -a v0.1.2`) to mark the stage of the developments even in my research repositories and then `git describe` gives a nice human readable representation of the HEAD. I think it would be nice if this was used in `tagsave` too.

For example if `git describe` spits out this: `12.0.0-alpha.1-8-g6f60bd1d-D`, I immediately know that it's 8 commits ahead of version 12.0.0-alpha.1, it's a dirty repository (`-D`) and the last commit is `6f60bd1d`.

My PR just falls back to the commit hash if there is no annotated tags and also preserves the `_dirty` suffix. Of course now the `current_commit()` is not really referring to the `current_commit`, so the naming should be discussed also.

What do you think?

### CODE REVIEW

**Comment by tamasgal** [comment]
I did not manage to reproduce the errors of the CI on my own machine yet 🤔

**Comment by Datseris** [comment]
Thank you very much for the contribution! I think I like this change and I will review properly once I have some extra free time in a couple of days (currently correcting my thesis :P )

**Comment by tamasgal** [comment]
> Thank you very much for the contribution! I think I like this change and I will review properly once I have some extra free time in a couple of days (currently correcting my thesis :P )

Sure, no hurries ;) I am still writing mine (a few weeks left) 😬

File: src/saving_tools.jl [hunk 1]
```diff
@@ -19,24 +19,35 @@ julia> current_commit()
...
-        @warn "The Git repository is dirty! Adding appropriate comment to "*
-        "commit id..."
-        return c*"_dirty"
+        suffix = "_dirty"
+        @warn "The Git repository is dirty! Appending $(suffix) to the commit ID"
+    end
+    # then we return the output of `git describe` or the latest commit hash
+    # if no annotated tags are available
+    repo = LibGit2.GitRepo(gitpath)
+    c = begin try
```
Comment (Datseris) [inline src/saving_tools.jl]: what does the `begin` keyword achieve here?

File: src/saving_tools.jl [hunk 2]
```diff
@@ -19,24 +19,35 @@ julia> current_commit()
...
+julia> current_commit(another_path_with_annotated_git_tags)
+"v1.2.3-g7364ab"
 ```
 """
 function current_commit(gitpath = projectdir())
     # Here we test if the gitpath is a git repository.
     try
         repo = LibGit2.GitRepo(gitpath)
     catch er
-        @warn "The current project directory is not a Git repository, "*
```
Comment (Datseris) [inline src/saving_tools.jl]: Here `*` is necessary. If you remove it the string on the line below is not chained to the above and only the first line is displayed as a warning.

File: src/saving_tools.jl (hunk 1 above)
Comment (tamasgal) [reply src/saving_tools.jl]: Ah, that's a leftover, I'll remove it!

File: src/saving_tools.jl (hunk 2 above)
Comment (tamasgal) [reply src/saving_tools.jl]: Yep! That was a typo... thanks!

**Comment by tamasgal** [comment]
> * are the 7-string shortcut strings always equivalent to a commit? From them can you always find the correct commit? (because then I am confused as to why do commit IDs have so much more characters)

As far as I know the abbreviations are guaranteed to be safe, as git will check for possible clashes. Once a clash is detected, the number of characters will increase.
The same holds for all kind of commit hash usages in git (like `git checkout abcd...`) where you can only provide a few characters and git will not complain if it's unique.

> * Indeed the name does not make sense anymore. I propose `gitdescribe` (please deprecate it). But then, in the case where a tag doesn't exist, should we still use the string with 100 characters instead of just 6?

OK `gitdescribe` sounds good to me. I think falling back to the full commit hash is reasonable...

> * `tagsave`, etc. should indeed use `gitdescribe`, provided that it is documented and the way the string works is explained (like e.g. the amount of commits after the tag, etc.)

We can link to the official docs to `git describe`: https://git-scm.com/docs/git-describe and summarise it in a few words I guess?

> * Please add a changelog entry / version update (0.7.0) / docstring

:+1: 

> * Please test the new functions in a case where a commit exists, but no tags. Currently in the tests `current_commit` is tested based on the repo of `DrWatson` that does have a tag. Maybe a quick `initialize_project` and call `gitdescribe` on this project?

For sure :+1:

It will take some time since I am also very busy with my PhD

**Comment by Datseris** [comment]
Alright, I agree with everything you said.

> It will take some time since I am also very busy with my PhD

No problemo, (I just finished the corrections to mine, hooray), just gimme a tag when you think I should merge!

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,10 @@
 # 0.7.0
 * New macro `@savename` that is a shortcut for `savename(@dict vars...)`
+* New function `gitdiscribe` (see below)
+* **[DEPRECATED]** `current_commit()` has been deprecated and replaced by
```
Comment (tamasgal) [inline CHANGELOG.md]: `describe` vs. `discribe`

**Comment by tamasgal** [comment]
Ah, thanks for the follow-ups 😃

**Pull request merged**